core/src/main/scala/kafka/coordinator/DelayedJoin.scala

```diff
@@ -31,14 +31,29 @@ import kafka.server.DelayedOperation
  */
 private[coordinator] class DelayedJoin(coordinator: GroupCoordinator,
                                        group: GroupMetadata,
-                                       rebalanceTimeout: Long)
-  extends DelayedOperation(rebalanceTimeout) {
+                                       delayMs: Int,
+                                       val rebalanceTimeOut: Long)
+  extends DelayedOperation(delayMs) {
 
   // overridden since tryComplete already synchronizes on the group. This makes it safe to
   // call purgatory operations while holding the group lock.
   override def safeTryComplete(): Boolean = tryComplete()
 
   override def tryComplete(): Boolean = coordinator.tryCompleteJoin(group, forceComplete)
-  override def onExpiration() = coordinator.onExpireJoin()
-  override def onComplete() = coordinator.onCompleteJoin(group)
+  override def onExpiration() =
+    coordinator.onExpireJoin()
+
+  override def onComplete() = {
+    group synchronized {
```

Comment:
@guozhangwang i updated this as if what i noticed is that if the `timerTask` in purgatory is executed at the same time it is being cancelled there is a race. So it might end up with this firing when the group has already rebalanced and is in state `AwaitingSync` in which case it causes more rebalances and it can end up not rebalancing the entire group. This, however resolves that issue. I'm not sure if there is a better way to achieve this?